lib/Targets/AIERTX.cpp

```diff
@@ -0,0 +1,457 @@
+    auto tileLoc =
+        XAie_TileLoc(muxOp.getTileOp().getCol(), muxOp.getTileOp().getRow());
+    Block &b = muxOp.getConnections().front();
+    for (auto connectOp : b.getOps<ConnectOp>()) {
+      // demux!
+      if (connectOp.getSourceBundle() == WireBundle::North)
+        TRY_XAIE_API_EMIT_ERROR(muxOp, XAie_EnableAieToShimDmaStrmPort,
+                                &devInst, tileLoc, connectOp.sourceIndex());
+      // mux
+      if (connectOp.getDestBundle() == WireBundle::North)
+        TRY_XAIE_API_EMIT_ERROR(muxOp, XAie_EnableShimDmaToAieStrmPort,
+                                &devInst, tileLoc, connectOp.destIndex());
+    }
+  }
+
+  for (auto switchboxOp : targetOp.getOps<ShimSwitchboxOp>()) {
+    Block &b = switchboxOp.getConnections().front();
+    auto tileLoc = XAie_TileLoc(switchboxOp.getCol(), 0);
+    for (auto connectOp : b.getOps<ConnectOp>())
+      TRY_XAIE_API_EMIT_ERROR(
+          switchboxOp, XAie_StrmConnCctEnable, &devInst, tileLoc,
+          WIRE_BUNDLE_TO_STRM_SW_PORT_TYPE.at(connectOp.getSourceBundle()),
+          connectOp.sourceIndex(),
+          WIRE_BUNDLE_TO_STRM_SW_PORT_TYPE.at(connectOp.getDestBundle()),
+          connectOp.destIndex());
+  }
+
+  // Cascade configuration
+  if (targetModel.getTargetArch() == AIEArch::AIE2) {
+    for (auto configOp : targetOp.getOps<ConfigureCascadeOp>()) {
+      TileOp tile = cast<TileOp>(configOp.getTile().getDefiningOp());
+      auto tileLoc = XAie_TileLoc(tile.getCol(), tile.getRow());
+      TRY_XAIE_API_EMIT_ERROR(
+          targetOp, XAie_CoreConfigAccumulatorControl, &devInst, tileLoc,
+          WIRE_BUNDLE_TO_STRM_SW_PORT_TYPE.at(
+              static_cast<WireBundle>(configOp.getInputDir())),
+          WIRE_BUNDLE_TO_STRM_SW_PORT_TYPE.at(
+              static_cast<WireBundle>(configOp.getOutputDir())));
+    }
+  }
+  return success();
+}
+
+LogicalResult AIERTXControl::enableCoresInDevice(DeviceOp &targetOp) {
+  // Start execution of all the cores.
+  for (auto tileOp : targetOp.getOps<TileOp>()) {
+    auto tileLoc = XAie_TileLoc(tileOp.colIndex(), tileOp.rowIndex());
+    if (!tileOp.isShimTile() && tileOp.getCoreOp())
+      TRY_XAIE_API_EMIT_ERROR(targetOp, XAie_CoreEnable, &devInst, tileLoc);
+  }
+  return success();
+}
+
+void AIERTXControl::dmaUpdateBdAddr(int col, int row, size_t addr,
+                                    size_t bdId) {
+  auto tileLoc = XAie_TileLoc(col, row);
+  TRY_XAIE_API_FATAL_ERROR(XAie_DmaUpdateBdAddr, &devInst, tileLoc, addr, bdId);
+}
+
+void AIERTXControl::startTransaction() {
+  TRY_XAIE_API_FATAL_ERROR(XAie_StartTransaction, &devInst,
+                           XAIE_TRANSACTION_DISABLE_AUTO_FLUSH);
+}
+
+void AIERTXControl::exportSerializedTransaction() {
+  XAie_TxnInst *txnInst = XAie_ExportTransactionInstance(&devInst);
+  std::ios_base::fmtflags f(std::cout.flags());
+  for (size_t i = 0; i < txnInst->NumCmds; ++i) {
+    std::cout.flags(f);
+    std::cout << "Txn OpCode: " << std::hex
```

Comment:
Since you are using LLVM, you could use the Pythonic `llvm::formatv` from
```c++
#include "llvm/Support/FormatVariadic.h"
```
instead of old `<<` spaghetti, while waiting for C++20 or compile with C++20/23...

Reply:
this is just for sanity checking - i'm going to ditch this soon (but thanks for the pointer to `format`.